### ISSUE TITLE
Feature request: Additional audio options (e.g. Sound fonts, GUS)

### ISSUE THREAD

**Issue by PatTheMav** [alunbestor/Boxer#47]
Adding to the amazing work done for MT-32 support in Boxer, I'd very much like to see a possibility to also configure additional audio options, e.g.:
- Selection of Sound fonts - Core Audio is able to use SF2 sound font files, DOSBOX supposedly supports this in OS X natively (see http://sourceforge.net/p/dosbox/patches/242/)
- Global configuration for Gravis Ultrasound - if given an ULTRASND folder, it would be awesome if this folder could be dropped on the config window just like the MT-32 ROMs which would enable GUS for all games and "mount" the folder properly (setting the "ultradir" config variable accordingly).

A bit of an explanation for this: Even though many games tout MT-32 support (and have appropriate drivers) not all of them have been composed using an MT-32 (or their MT-32 tracks haven't). Case in point: X-Wing's floppy version has a MT-32 sound track, it's General Midi sound track was converted from the MT-32 version. X-Wing's CD Rom version however sports only the General Midi tracks (even for the MT-32). TIE Fighter was composed on a Roland SC-55 which is a General Midi device, so it's MT-32 track is also just a converted General Midi soundtrack.

So while one might want to use MT-32 for - say - Monkey Island (which truly shines using it), TIE Fighter or Dark Forces will sound better using General Midi and a proper modern sound font.

As Boxer passes General Midi music to core audio (bypassing its internal MT-32 support), it could use an available sound font. Configuring this "out of the box" on OS X has become quite cumbersome ever since Quicktime X and requires outdated software (often requiring Rosetta) or doing a full install of Xcode to acquire needed programs.

Edit: Looking at the repository it seems that the underlying system is able to load a user-specified sound font, so the UI exposing that feature is "just" missing.

**Comment by lavadrop** [comment]
Just my 2¢: I have an MT32 and a SoundCanvas. The MT32 is automatically detected (thanks for the compliment :P ) but the SoundCanvas only works if the .conf file is altered to select coremidi as the sound server instead of the default coreaudio.

**Comment by PatTheMav** [comment]
Just FYI - switching the gs_Instruments.dls with alternative ones does not work anymore with OSX El Capitan in the new "rootless" mode, as all write access to the /System/ folder is prohibited.

Also I haven't found a way to make the system use a CoreAudio.component in an alternative folder (e.g. /Library/ yet.

**Comment by MaddTheSane** [comment]
To my knowledge, Boxer shouldn’t need to write to the gs_instrument.dls file. Read access, yes, but not write.

**Comment by PatTheMav** [comment]
> To my knowledge, Boxer shouldn’t need to write to the gs_instrument.dls file. Read access, yes, but not write.

You're absolutely right; this was more intended as an update to the "fix" @JoshuaPettus suggested which made Boxer games use Timbres of Heaven or FluidR3 as sound fonts through modifying the system's gs_instruments file (as "native" sound font support is not built into the currently released version).

**Comment by lavadrop** [comment]
This could be a significant issue for people using soft synths and not just
limited to Boxer, but all current Core Audio apps like GarageBand

**Comment by PatTheMav** [comment]
> This could be a significant issue for people using soft synths and not just
> limited to Boxer, but all current Core Audio apps like GarageBand

As far as I understood the "rootless" specifications, Apple signed installers are still able to write to these folders - like the rest of Apple's encroaching walled garden approach to OSX, you can either a) disable rootless yourself (at the cost of losing this "security feature") or shell out money for a developer certificate and sign your installers.

**Comment by Goi-Yaas-Dinn** [comment]
Dropbox links are dead.

**Comment by kode54** [comment]
Further alternative MIDI options: Audio Unit picker, to allow one to use things such as the extremely overpriced Roland Sound Canvas VA. It also has an internal API, but that shall not be discussed here, since it goes deeper than the level at which they implemented the processing-expensive software licensing system that happens to add over 10 seconds to the startup time.

**Comment by almeath** [comment]
I have tried pointing Boxer 2 alpha to a few different SF2 sound fonts using the specified terminal command. It generally works very well. My only problem is the volume of some fonts is way too loud and can sometimes drown out the voice and sound effects in various games. Some games allow for independent music volume control, but others just have one master volume control.

Is there a way to adjust the volume at which Boxer plays the MIDI music, independently of the master volume slider?

**Comment by Goi-Yaas-Dinn** [comment]
@almeath Dead project.

**Comment by almeath** [comment]
@Goi-Yaas-Dinn There are active discussions taking place on the issues tracker for this "dead project". We all know that Boxer has not been substantively updated by the developer for a number of years. It does not make my question any less valid. The app continues to work well in the current release of macOS. There is also an active fork being worked on. I was not expecting an answer from the developer, and if there is no solution available in Boxer's current state, then so be it.

**Comment by Goi-Yaas-Dinn** [comment]
> There is also an active fork being worked on.

@almeath ...Eh? Link? I know it's been forked several times, (including by an OpenEmu contributor), but was completely unaware any were actually "active".

**Comment by almeath** [comment]
I say "active" in the sense that they have been updated in the last couple of years since official development of Boxer stopped. This one has been updated in the last two weeks, incorporating updates from the DOSBox SVN. 

https://github.com/MaddTheSane/Boxer/tree/updateDosBox

I have not had time to test this out, but it just goes to show that there is continued interest in keeping Boxer alive as long as possible.

**Comment by almeath** [comment]
I used the _defaults write_ command to point Boxer 2 alpha to a custom sound font, and that works fine. However, when I use Boxer Bundler to create a standalone app (obviously using the Boxer 2 alpha code) I notice that the resulting apps do not recognize this re-pointing and they default to the built in DLS midisynth in macOS.

Is there a way to point these standalone Boxer Bundler apps to sound fonts, or does it only work with the main Boxer app itself?

**Comment by almeath** [comment]
I know I could convert an .sf2 sound font to .dls with the right software, and entirely replace my system level sound font. 

But that seems like overkill if there is a simpler way to do this.

**Comment by MaddTheSane** [comment]
I am fairly certain CoreAudio can use both `.sf2` and `.dls` sound fonts. The sample code in [this](https://developer.apple.com/library/content/samplecode/PlaySequence/Introduction/Intro.html) project should help, particularly the `-b` flag.

**Comment by almeath** [comment]
Yes, I think there has always been an ability in macOS to override the in-built GM instrument set with sound fonts (either sf2 or DLS) but it needs to be set on a per-application basis, such as within QuickTime or GarageBand etc. I cannot find a way in macOS High Sierra to force the use of an SF2 sound font system-wide across all apps, otherwise my problem would be solved.

Alun provided the terminal command to point Boxer 2 alpha to a corresponding sound font and override the default core audio instruments. SVN DOSBox achieves the same thing by using the core audio or core midi setting and pointing to the soundfont in the configuration file. 

However, for various reasons, I want to use standalone Boxer applications constructed with Boxer Bundler. These standalone apps ignore the terminal command used with the main Boxer application, so the individual Boxer apps fall back on the in-built GM instruments. Unlike DOSBox, I cannot manually point to the sound font in the configuration file.

So, I assumed that unless these standalone apps can be made to follow the re-direct command noted by Alun, then the only way I can see to force the use of the preferred sound font is to replace the GM instruments at the system level. 

I also thought of using FluidSynth and having the MIDI output redirected through that, similar to when using MUNT, but there are no decent FluidSynth setups for newer versions of macOS of which I am aware.

Anyway, I’ll keep researching this.

**Comment by kode54** [comment]
In case anyone wants to look at my Gists, there's a SoundFont List loader, which supports both flat file lists (v1), and JSON based lists (v2), and nesting the lists. It currently is designed to auto load SoundFonts with BASSMIDI, but it could be adapted to load with FluidSynth, or any other synthesizer that supports multiple SoundFont (or DLS) banks per instance, and also picking out individual presets from these banks, and setting a volume override on them. It also has a feature that depends on the synthesizer, which can load presets for specific MIDI channels, which just requires the synthesizer to support that somehow. For BASSMIDI, I implement that using the bank LSB control.

E: Ah, here it is: https://gist.github.com/kode54/a7bb01a0db3f2e996145b77f0ca510d5
